core/components/menu/menu.js

```diff
@@ -0,0 +1,543 @@
+  Blockly.unbindEvent_(this.mouseEnterHandler_);
+  Blockly.unbindEvent_(this.mouseLeaveHandler_);
+  Blockly.unbindEvent_(this.onKeyDownWrapper_);
+};
+
+// Child component management.
+
+/**
+ * Map of DOM IDs to child controls.  Each key is the DOM ID of a child
+ * control's root element; each value is a reference to the child control
+ * itself.  Used for looking up the child control corresponding to a DOM
+ * node in O(1) time.
+ * @type {?Object}
+ * @private
+ */
+Blockly.Menu.prototype.childElementIdMap_ = null;
+
+/**
+ * Creates a DOM ID for the child control and registers it to an internal
+ * hash table to be able to find it fast by id.
+ * @param {Blockly.Component} child The child control. Its root element has
+ *     to be created yet.
+ * @private
+ */
+Blockly.Menu.prototype.registerChildId_ = function(child) {
+  // Map the DOM ID of the control's root element to the control itself.
+  var childElem = child.getElement();
+
+  // If the control's root element doesn't have a DOM ID assign one.
+  var id = childElem.id || (childElem.id = child.getId());
+
+  // Lazily create the child element ID map on first use.
+  if (!this.childElementIdMap_) {
+    this.childElementIdMap_ = {};
+  }
+  this.childElementIdMap_[id] = child;
+};
+
+/**
+ * Returns the child menu item that owns the given DOM node, or null if no such
+ * menu item is found.
+ * @param {Node} node DOM node whose owner is to be returned.
+ * @return {?Blockly.MenuItem} Menu Item for which the DOM node belongs to.
+ * @protected
+ */
+Blockly.Menu.prototype.getMenuItem = function(node) {
+  // Ensure that this menu actually has child controls before
+  // looking up the menu item.
+  if (this.childElementIdMap_) {
+    var elem = this.getElement();
+    while (node && node !== elem) {
+      var id = node.id;
+      if (id in this.childElementIdMap_) {
+        return this.childElementIdMap_[id];
+      }
+      node = node.parentNode;
+    }
+  }
+  return null;
+};
+
+/** @override */
+Blockly.Menu.prototype.addChildAt = function(control, index, opt_render) {
+
+  // Disable mouse event handling by child controls.
+  // control.setHandleMouseEvents(false);
+
+  var srcIndex =
+      (control.getParent() == this) ? this.indexOfChild(control) : -1;
+
+  // Let the superclass implementation do the work.
+  Blockly.Menu.superClass_.addChildAt.call(
+      this, control, index, opt_render);
+
+  if (control.isInDocument() && this.isInDocument()) {
+    this.registerChildId_(control);
+  }
+
+  this.updateHighlightedIndex_(srcIndex, index);
+};
+
+// Highlight management.
+
+/**
+ * Updates the highlighted index when children are added or moved.
+ * @param {number} fromIndex Index of the child before it was moved, or -1 if
+ *     the child was added.
+ * @param {number} toIndex Index of the child after it was moved or added.
+ * @private
+ */
+Blockly.Menu.prototype.updateHighlightedIndex_ = function(
+    fromIndex, toIndex) {
```

Comment:
I don't understand this function.  Will get an explanation in person.

Reply:
This was to support dynamic adding of menu item elements. Since we don't need that built in, I'll remove for now.

Reply:
I should clarify, that this is while the menu item is open. I know we have a dynamic dropdown in the playground, but that the menu there is created every time the dropdown is open.